### PR TITLE
feat(v2): prompt user when default port is in use

### PR DESCRIPTION
**Pull request by taylorallen0913**
## Motivation

This PR implements what was proposed in #3005.

When a user already has a process running on the default docusaurus port, docusaurus finds the next open port and runs the client on it. Based on a simular implementation in create-react-app, this feature warns the user if something is already running on the default port, and asks them if they want to run the client on a different port.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/facebook/docusaurus/blob/master/CONTRIBUTING.md#pull-requests)?

yes

## Test Plan

Run a program on port 3000 (I ran create-react-app)

`yarn start`

When you try to run `yarn start` with another app already running on the Docusaurus default port:

docusaurus

When you select "Yes":

docusaurus2

When you select "No":

docusaurus3

## Related PRs

(If this PR adds or changes functionality, please take some time to update the docs at https://github.com/facebook/docusaurus, and link to your PR here.)

### CODE REVIEW

File: packages/docusaurus/src/webpack/getProcessForPort.ts [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+// Gets process id of what is on port
+function getProcessIdOnPort(port: number): string {
+  return execSync(`lsof -i:${port} -P -t -sTCP:LISTEN`, execOptions)
+    .toString()
+    .split('\n')[0]
+    .trim();
+}
+
+// Gets package name in a directory
+async function getPackageNameInDirectory(
```
Comment (slorber) [inline packages/docusaurus/src/webpack/getProcessForPort.ts:35]: do need to export it afaik

**Comment by slorber** [comment]
@yangshun I guess it's not a problem to copy facebook code from one project to another right?

This is heavily inspired from current CRA code: https://github.com/facebook/create-react-app/blob/f5c3bdb65480f93b2d4a4c2f3214fc50753de434/packages/react-dev-utils/WebpackDevServerUtils.js#L448

File: packages/docusaurus/src/commands/start.ts [hunk 2]
```diff
@@ -11,7 +11,6 @@ import chokidar from 'chokidar';
 import express from 'express';
 import HtmlWebpackPlugin from 'html-webpack-plugin';
 import path from 'path';
-import portfinder from 'portfinder';
```
Comment (slorber) [inline packages/docusaurus/src/commands/start.ts:14]: Remove the dependency from package, this seems unused now

File: packages/docusaurus/src/webpack/getProcessForPort.ts [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+    'pipe', // stdin (default)
+    'pipe', // stdout (default)
+    'ignore', // stderr
+  ],
+};
+
+// Checks if process is a React app
+function isProcessAReactApp(processCommand) {
+  return /^node .*react-scripts\/scripts\/start\.js\s?$/.test(processCommand);
+}
```
Comment (slorber) [inline packages/docusaurus/src/webpack/getProcessForPort.ts:24]: that seems CRA specific and useless here

File: packages/docusaurus/src/webpack/utils.ts [hunk 4]
```diff
@@ -148,3 +156,53 @@ export function applyConfigureWebpack(
   }
   return config;
 }
+
+export async function choosePort(
```
Comment (slorber) [inline packages/docusaurus/src/webpack/utils.ts:160]: what about putting everything in a choosePort.ts file? I'd like to have all this feature in a single place (unlike CRA), and maybe comment that this code is inspired from CRA codebase would be useful

File: packages/docusaurus/src/webpack/getProcessForPort.ts (hunk 1 above)
Comment (taylorallen0913) [reply packages/docusaurus/src/webpack/getProcessForPort.ts:35]: Are you saying to remove the `export default function getProcessForPort` ? I am using the function in another file.

**Pull request merged**